File: god-sim-prototype/src/powers.js
/* SLATE Phase 0 — acts of god. Three powers, each a perturbation the
   living world will interpret in its own time. Every act is chronicled;
   nothing a god does is forgotten. */
(function (S) {
  'use strict';

  function seedGold(w, x, y) {
    x = Math.round(x); y = Math.round(y);
    if (!w.inB(x, y)) return { ok: false, msg: 'Beyond the edge of the map.' };
    const b = w.biome[w.idx(x, y)];
    if (b !== S.B.MOUNTAIN && b !== S.B.HILLS) {
      return { ok: false, msg: 'The vein must be seeded in high stone — choose mountains or hills.' };
    }
    const v = { x, y, revealed: true, divine: true };
    w.veins.push(v);
    S.chronicle.add(w, 'goldSeed', { region: w.regionNameAt(x, y, 'the high stone'), x, y });
    w.dirty.features = true;
    return { ok: true, x, y };
  }

  function curseWeather(w, x, y) {
    x = Math.round(x); y = Math.round(y);
    if (!w.inB(x, y)) return { ok: false, msg: 'Beyond the edge of the map.' };
    let anyLand = false;
    for (let dy = -3; dy <= 3 && !anyLand; dy++) for (let dx = -3; dx <= 3 && !anyLand; dx++) {
      if (w.isLandAt(x + dx, y + dy)) anyLand = true;
    }
    if (!anyLand) return { ok: false, msg: 'Curse the living land — the open sea cares nothing for weather.' };
    if (w.stormAt(x, y)) return { ok: false, msg: 'That sky is already yours.' };
    w.storms.push({ x, y, r: 5, start: w.tick });
    w.recomputeFert();
    const near = w.settlementNear(x, y, 8);
    S.chronicle.add(w, 'curse', { place: near ? near.name : w.regionNameAt(x, y, 'the open country'), x, y });
    w.dirty.features = true;
    return { ok: true, x, y };
  }

  function blessLand(w, x, y) {
    x = Math.round(x); y = Math.round(y);
    if (!w.inB(x, y)) return { ok: false, msg: 'Beyond the edge of the map.' };
    if (!w.isLandAt(x, y)) return { ok: false, msg: 'Bless the soil, not the sea.' };
    if (w.stormAt(x, y)) return { ok: false, msg: 'The storm would devour the blessing — this sky is cursed.' };
    if (w.blessAt(x, y)) return { ok: false, msg: 'This land already carries your favor.' };
    w.blesses.push({ x, y, r: 5, start: w.tick });
    w.recomputeFert();
    const near = w.settlementNear(x, y, 8);
    S.chronicle.add(w, 'bless', { place: near ? near.name : w.regionNameAt(x, y, 'the open country'), x, y });
    w.dirty.features = true;
    return { ok: true, x, y };
  }

  S.powers = { seedGold, curseWeather, blessLand };
})(typeof SLATE !== 'undefined' ? SLATE : (globalThis.SLATE = {}));

File: SlateSim/Runtime/Powers.cs
// SLATE — acts of god. Port of god-sim-prototype/src/powers.js.
// Three powers, each a perturbation the living world will interpret in its
// own time. Every act is chronicled; nothing a god does is forgotten.
using System;

namespace Slate.Sim
{
    public struct PowerResult
    {
        public bool Ok;
        public string Msg;
        public int X, Y;
    }

    public static class Powers
    {
        // JS Math.round rounds half toward +infinity.
        private static int RoundJs(double v) => (int)Math.Floor(v + 0.5);

        public static PowerResult SeedGold(World w, double xIn, double yIn)
        {
            int x = RoundJs(xIn), y = RoundJs(yIn);
            if (!w.InB(x, y)) return new PowerResult { Ok = false, Msg = "Beyond the edge of the map." };
            byte b = w.Biome[w.Idx(x, y)];
            if (b != B.MOUNTAIN && b != B.HILLS)
                return new PowerResult { Ok = false, Msg = "The vein must be seeded in high stone — choose mountains or hills." };
            w.Veins.Add(new Vein { X = x, Y = y, Revealed = true, Divine = true });
            w.Chronicle.Add(w, "goldSeed", new EvData { Region = w.RegionNameAt(x, y, "the high stone"), X = x, Y = y });
            w.Dirty.Features = true;
            return new PowerResult { Ok = true, X = x, Y = y };
        }

        public static PowerResult CurseWeather(World w, double xIn, double yIn)
        {
            int x = RoundJs(xIn), y = RoundJs(yIn);
            if (!w.InB(x, y)) return new PowerResult { Ok = false, Msg = "Beyond the edge of the map." };
            bool anyLand = false;
            for (int dy = -3; dy <= 3 && !anyLand; dy++)
                for (int dx = -3; dx <= 3 && !anyLand; dx++)
                    if (w.IsLandAt(x + dx, y + dy)) anyLand = true;
            if (!anyLand) return new PowerResult { Ok = false, Msg = "Curse the living land — the open sea cares nothing for weather." };
            if (w.StormAt(x, y) != null) return new PowerResult { Ok = false, Msg = "That sky is already yours." };
            w.Storms.Add(new Zone { X = x, Y = y, R = 5, Start = w.Tick });
            w.RecomputeFert();
            var near = w.SettlementNear(x, y, 8);
            w.Chronicle.Add(w, "curse", new EvData { Place = near != null ? near.Name : w.RegionNameAt(x, y, "the open country"), X = x, Y = y });
            w.Dirty.Features = true;
            return new PowerResult { Ok = true, X = x, Y = y };
        }

        public static PowerResult BlessLand(World w, double xIn, double yIn)
        {
            int x = RoundJs(xIn), y = RoundJs(yIn);
            if (!w.InB(x, y)) return new PowerResult { Ok = false, Msg = "Beyond the edge of the map." };
            if (!w.IsLandAt(x, y)) return new PowerResult { Ok = false, Msg = "Bless the soil, not the sea." };
            if (w.StormAt(x, y) != null) return new PowerResult { Ok = false, Msg = "The storm would devour the blessing — this sky is cursed." };
            if (w.BlessAt(x, y) != null) return new PowerResult { Ok = false, Msg = "This land already carries your favor." };
            w.Blesses.Add(new Zone { X = x, Y = y, R = 5, Start = w.Tick });
            w.RecomputeFert();
            var near = w.SettlementNear(x, y, 8);
            w.Chronicle.Add(w, "bless", new EvData { Place = near != null ? near.Name : w.RegionNameAt(x, y, "the open country"), X = x, Y = y });
            w.Dirty.Features = true;
            return new PowerResult { Ok = true, X = x, Y = y };
        }
    }
}
